### ISSUE TITLE
Can I do any action just after return my response with an API GraphQL

### ISSUE THREAD

**Issue by bruno1407** [graphql-dotnet/graphql-dotnet#2647]
-->

## Summary

I am trying to perform an action right after returning my response to a GraphQL API request. Let me explain: I would like to develop a GraphQL server that receives a certain request from another server and sends a request to this other server just after answering the first request

In comparison, it looks like the OnActionExecuted function with ActionFilterAttribute but I can't implement it for GraphQl, maybe it works only for Rest APIs

I also found the function : public async override Task AfterExecutionAsync(IExecutionContext context) from class GraphQL.Execution.DocumentExecutionListenerBase but I can't figure out how to implement it so that it triggers after returning my answer
https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/DocumentExecutionListener.cs

this is the idea but of course it is not what i have tried haha

```
public async Task<sendLinkRequest> sendLinkStatusUpdate(newLinkStatusInput input)
        {             
            return _sendLinkRequestService.sendLinkStatusUpdate(input);

            //Doing this after my function return the result
            var graphQLClient = new GraphQLHttpClient("https://exemple.com", new NewtonsoftJsonSerializer());
            var sendLinkStatusUpdate = new GraphQLRequest
            {
                Query = "mutation{sendLinkStatusUpdate(input : {newLinkStatus: \"active\"}) {errors {scope message}}}",
                Variables = new
                {
                    newLinkStatus = "active"
                }
            };
                
            var graphQLResponse = await graphQLClient.SendQueryAsync<sendLinkRequest>(sendLinkStatusUpdate);
            
        }
```
Anyway, if someone has an idea or a lead I'm interested

<!--
  To make it easier for us to help you, please include as much useful information as possible.

  Useful Links:
  - Documentation: https://graphql-dotnet.github.io/

  GraphQL.NET has community support channels, try asking your question on:

  - Gitter: https://gitter.im/graphql-dotnet/graphql-dotnet
  - Stack Overflow: https://stackoverflow.com/questions/tagged/graphql-dotnet

  Before opening a new issue, please search existing issues https://github.com/graphql-dotnet/graphql-dotnet/issues

**Comment by bruno1407** [comment]
Thanks for your comment ! I managed to do it like that :)
```

   public sendLinkRequest sendLinkStatusUpdate(newLinkStatusInput input)
    {
        TrackingActionFilter ta = new TrackingActionFilter();
        ta.ProcessCompleted += ta_nextAction; // register with an event
        sendLinkRequest test = ta.StartProcess(input, _sendLinkRequestService);

        return test;

    }

    public async static void ta_nextAction(object sender, EventArgs e)
    {
        var graphQLClient = new GraphQLHttpClient("https://talentusnewdemo.eu-west-1.beepleapp.eu/b/multi-tenant-api", new NewtonsoftJsonSerializer());
        var sendLinkStatusUpdate = new GraphQLRequest
        {
            Query = "mutation{sendLinkStatusUpdate(input : {newLinkStatus: \"active\"}) {errors {scope message}}}",
            Variables = new
            {
                newLinkStatus = "active"
            }
        };

        var graphQLResponse = await graphQLClient.SendQueryAsync<sendLinkRequest>(sendLinkStatusUpdate);
        Console.WriteLine("Process Completed!");
    }

 public class TrackingActionFilter
{
    public event EventHandler ProcessCompleted;

    public sendLinkRequest StartProcess(newLinkStatusInput input, IsendLinkRequestService _sendLinkRequestService)
    {
        Console.WriteLine("Process Started!");
       
        OnProcessCompleted(EventArgs.Empty); //No event data
        return _sendLinkRequestService.sendLinkStatusUpdate(input);
    }

    protected virtual void OnProcessCompleted(EventArgs e)
    {
        ProcessCompleted?.Invoke(this, e);
    }
}

```

**Comment by Shane32** [comment]
I'd be thinking along these lines:

1. Does it need to execute specifically AFTER the response?  Why so?  How would it know if the response has been delivered successfully or completely?  Assuming it was delivered successfully, wouldn't it need to be processed as well?  How would it know if it had been processed?
2. Assuming that it does not exactly need to be delivered after the response, how about a `async void` task that runs in parallel to returning the data from the query
3. Does it need data from the query result in order to send this other request?  If so, does the `AfterExecutionAsync` method parameters contain the necessary information?  Or perhaps it needs the actual result object.  Or perhaps the json-formatted result text.
4. If `AfterExecutionAsync` does not have enough information, you'd probably need the result of the document executer's ExecuteAsync method
5. If the actual JSON text is needed, you'd need to serialize the result to a JSON string rather than (or in addition to) writing it directly to the stream
6. If you need it to actually send the response before making the other call, you'd probably need to call Response.Close to flush and close the response network stream before starting the other call
7. If you use `async void` you need to be keenly aware not to use the http context or any scoped objects obtained from dependency injection within the method.  Rather, call `CreateScope` on any instance of `IServiceProvider` to create a dedicated service scope, and then use `GetRequiredService` on the created scope to pull fresh instances from DI.  It may be helpful to use `static async void` in the method declaration.  If you want to continue to use DI rather than `GetRequiredService` for various dependencies, you can first encapsulate your code in a class with any necessary dependencies, then within the `static async void` method, you call `CreateScope` and `ActivatorUtilities.GetOrCreateService` to create an instance of the class, and then call a method like `RunAsync` on it.  The class can be private to the encapsulating class and does not need to be registered within the DI framework.

**Comment by bruno1407** [comment]
thank you very much for this very complete answer and its good reflections to have. 
Hi @Shane32 

Indeed, the code that I tested above, in fact, does not work for what I want to do. 

I would like to be able to send first the answer of the first queries before sending the second one, if possible with a small delay of one second between the two. 

The answer of the first mutation will allow the server with which I communicate to be able to listen to my second mutation

So, if we follow your directions, you would advise me to follow option 6 ?
6: If you need it to actually send the response before making the other call, you'd probably need to call Response.Close to flush and close the response network stream before starting the other call

**Comment by Shane32** [comment]
That’s probably what I would try first. Keep in mind that there are also system buffers that we cannot control or monitor. But hopefully with an adequate delay it will work for your needs.

**Comment by sungam3r** [comment]
@bruno1407 Did you manage to solve this issue?

**Comment by sungam3r** [comment]
Closed due to inactivity.
